server/artifacts/artifact_server.go

```diff
@@ -344,14 +344,42 @@ func (a *ArtifactServer) httpBadRequestError(w http.ResponseWriter) {
 }
 
 func (a *ArtifactServer) httpFromError(err error, w http.ResponseWriter) {
+	if err == nil {
+		return
+	}
+	statusCode := http.StatusInternalServerError
 	e := &apierr.StatusError{}
 	if errors.As(err, &e) {
 		// There is a http error code somewhere in the error stack
-		statusCode := int(e.Status().Code)
-		http.Error(w, http.StatusText(statusCode), statusCode)
+		statusCode = int(e.Status().Code)
 	} else {
-		// Unknown error - return internal error
-		a.serverInternalError(err, w)
+		switch resolvedError := err.(type) {
+		case argoerrors.ArgoError:
+			switch resolvedError.Code() {
+			case argoerrors.CodeUnauthorized:
+				statusCode = http.StatusUnauthorized
+			case argoerrors.CodeForbidden:
+				statusCode = http.StatusForbidden
+			case argoerrors.CodeNotFound:
+				statusCode = http.StatusNotFound
+			case argoerrors.CodeBadRequest:
+				statusCode = http.StatusBadRequest
+			case argoerrors.CodeNotImplemented:
+				statusCode = http.StatusNotImplemented
+			case argoerrors.CodeTimeout, argoerrors.CodeInternal:
+				statusCode = http.StatusInternalServerError
+			default:
+				statusCode = http.StatusInternalServerError
+			}
```

Comment:
while I could have excluded argoerrors.CodeTimeout and argoerrors.CodeInternal and let them fall through to default, I left them in to show they were addressed